feat: Rest API interface for filter service.

NagyZoltanPeter opened the pull request:
# Description
<!--- Describe your changes to provide context for reviewrs -->
Draft PR for preliminary review, I appreciate any comment on it. Specially on the test_rest_filter.nim tests as 
peers and not finding each other. Probably overlooked something in setting up.

- Interface added
- tests are added but not yet functional. 

# Changes

<!-- List of detailed changes -->
- [x] Filter rest api is added
- [x] Tests are added for sub/unsub basic functionality.

## How to test
nim c -r -d:chronicles_log_level=DEBUG --verbosity=0 --hints=off --outdir=build ./tests/v2/wakunode_rest/test_rest_filter.nim test "Subscribe a node to an array of topics - POST /filter/v1/subscriptions"

tracking: [#1851](https://github.com/waku-org/nwaku/issues/1851)

jm-clius reviewed (commented):
Thanks for this. Functionally, this looks good. Two comments:

1. I apologise that this wasn't clear enough in https://github.com/waku-org/nwaku/issues/1851, but we want this effort to also consider the "new" filter client API. This still follows the same "legacy" filter API functions as the existing JSON-RPC API for filter. We've since significantly improved the filter specification and the client API now defines different functions as documented in https://github.com/waku-org/nwaku/blob/master/waku/v2/waku_filter_v2/client.nim#L74-L106 and https://rfc.vac.dev/spec/12/ (you can still view the old version if you click on the link at the top of the RFC). These client API functions are not yet integrated in the `node` object, so we'll have to first expand the node to set up such a new filter client (in a next PR) and then expose API functions for these as well. Will provide more details in a 1:1. For now I propose, rough roadmap:
- get this version merged and tests fixed/integrated
- integrate "new" filter client in node and expose client functions in node API
- expose new filter client API functions in REST API

2. Could you provide more details as to what you see in terms of the tests failing? Perhaps some logs? Are all tests failing? We don't expect a filter client to connect to its service peer on startup - it makes ad-hoc connections when needed (i.e. only when it posts the subscription request). Perhaps you just need to add a `sleep` to allow the client to connect to the service node after `filterPost` or other action that requires the connection to be setup first?

NagyZoltanPeter commented:
Update:

Tests refactored a bit and now works ok. 
Basic functionality tested: subscribe/unsubscribe (original way, copying jsonrpc)

Branch squashed and rebased.

NagyZoltanPeter commented:
Update:

Tests refactored a bit and now works ok. 
Basic functionality tested: subscribe/unsubscribe (original way, copying jsonrpc)

Branch squashed and rebased.

NagyZoltanPeter commented:
Hi Hanno,
Yes it was not clear for me. Looking at the code and the existing openapi spec it seemed just an interface change.
I agree with the proposed steps and yet we need to discuss through due some unclear points remained even after more careful read of rfc. I have questions to the interface as well.
The latter, test issues are solved with help of Ivan already.


> Thanks for this. Functionally, this looks good. Two comments:
> 
> 1. I apologise that this wasn't clear enough in [feat: HTTP REST API: Filter  #1851](https://github.com/waku-org/nwaku/issues/1851), but we want this effort to also consider the "new" filter client API. This still follows the same "legacy" filter API functions as the existing JSON-RPC API for filter. We've since significantly improved the filter specification and the client API now defines different functions as documented in https://github.com/waku-org/nwaku/blob/master/waku/v2/waku_filter_v2/client.nim#L74-L106 and https://rfc.vac.dev/spec/12/ (you can still view the old version if you click on the link at the top of the RFC). These client API functions are not yet integrated in the `node` object, so we'll have to first expand the node to set up such a new filter client (in a next PR) and then expose API functions for these as well. Will provide more details in a 1:1. For now I propose, rough roadmap:
> 
> * get this version merged and tests fixed/integrated
> * integrate "new" filter client in node and expose client functions in node API
> * expose new filter client API functions in REST API
> 
> 2. Could you provide more details as to what you see in terms of the tests failing? Perhaps some logs? Are all tests failing? We don't expect a filter client to connect to its service peer on startup - it makes ad-hoc connections when needed (i.e. only when it posts the subscription request). Perhaps you just need to add a `sleep` to allow the client to connect to the service node after `filterPost` or other action that requires the connection to be setup first?

SionoiS reviewed (approved):
LGTM!

vpavlin commented on apps/wakunode2/app.nim:
```diff
@@ -548,6 +549,11 @@ proc startRestServer(app: App, address: ValidIpAddress, port: Port, conf: WakuNo
 
+  ## Filter REST API
+  if conf.filter:
+    let filterCache = rest_filter_api.MessageCache.init(capacity=30)
```
Should this be configurable like for `restRelayCacheCapacity` above, or at least defined in a constant? (I saw there are other "magic numbers" used below this when starting RPC server, but still:)

NagyZoltanPeter replied:
You're absolutely right about it. I will quick fix it, also I need to sign the commit that I had not set up earlier though

NagyZoltanPeter commented:
Fixed magic number, first not with conf but with global const. Let's see in v2 if we need a config for that value.

I had to come over some hurdles with signing the commit and I had to renew my ssh creds too...

vpavlin reviewed (approved):
LGTM

jm-clius commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+    if peerOpt.isNone():
+      raise newException(ValueError, "no suitable remote filter peers")
+
+    let handler: FilterPushHandler = proc(pubsubTopic: PubsubTopic, msg: WakuMessage) {.async, gcsafe, closure.} =
+        cache.addMessage(msg.contentTopic, msg)
```
Minor note, but any reason why we'd create a new `FilterPushHandler` for every new `post`? Wouldn't it be possible to create it outside the main method body and reuse the same one?

jm-clius reviewed (approved):
This LGTM! Well done on a thorough first PR. Minor comment/question below, but afaict it can be merged as-is.

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+    let resp = RestApiResponse.jsonResponse(data, status=Http200)
+    if resp.isErr():
+      debug "An error ocurred while building the json respose", error=resp.error
+      return RestApiResponse.internalServerError()
```
Shall we bring better feedback to the client?

```suggestion
     msg = "...."
     error msg
     return RestApiResponse.internalServerError(msg)
```

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+
+  let reqResult = decodeFromJsonBytes(T, reqBodyData)
+  if reqResult.isErr():
+    return err(RestApiResponse.badRequest())
```
Little nitpick to help debug possible error:

```suggestion
    return err(RestApiResponse.badRequest(reqResult.error))
```

Ivansete-status commented on waku/v2/node/rest/filter/types.nim:
```diff
@@ -0,0 +1,157 @@
+      unrecognizedFieldWarning()
+
+  if payload.isNone():
+    reader.raiseUnexpectedValue("Field `payload` is missing")
```
In other comments, I mentioned that we should avoid raising exceptions. Is fine for now in the case of the `types.nim` as we are doing the same in another protocols :)

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+func decodeRequestBody[T](contentBody: Option[ContentBody]) : Result[T, RestApiResponse] =
+  # Check the request body
+  if contentBody.isNone():
+    return err(RestApiResponse.badRequest())
```
Shall we have more feedback in case of error?

```suggestion
    return err(RestApiResponse.badRequest("content body is none"))
```

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+    let decodedBody = decodeRequestBody[FilterSubscriptionsRequest](contentBody)
+
+    if decodedBody.isErr():
+      error "Failed to decode body", error=decodedBody.error()
+      return decodedBody.error()
```
Shall we have better feedback in case of error?

```suggestion
      let msg = "Failed to decode body: " & decodedBody.error()
      error msg
      return RestApiResponse.badRequest(msg)
```

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+    let peerOpt = node.peerManager.selectPeer(WakuFilterCodec)
+
+    if peerOpt.isNone():
+      raise newException(ValueError, "no suitable remote filter peers")
```
I wouldn't raise Exception in this case. Instead, log the error and 

`return RestApiResponse.badRequest( msg )`

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+
+    let unsubFut = node.unsubscribe(req.pubsubTopic, req.contentFilters)
+    if not await unsubFut.withTimeout(futTimeout):
+      raise newException(ValueError, "Failed to unsubscribe from contentFilters")
```
```suggestion
      return RestApiResponse.internalServerError("Failed to unsubscribe from contentFilters")
```

Ivansete-status commented on waku/v2/node/rest/filter/handlers.nim:
```diff
@@ -0,0 +1,151 @@
+    # debug "get_waku_v2_filter_v1_messages", contentTopic=contentTopic
+
+    if contentTopic.isErr():
+      return RestApiResponse.badRequest()
```
In this proc, all the possible errors should be retrieved back to the client

```suggestion
      return RestApiResponse.badRequest("Error in content topic: " & contentTopic.error)
```

Ivansete-status reviewed (commented):
Thanks for that! you made it super fast and very well!
I add small details to mostly enhance the feedback in case of error.
Great job!

NagyZoltanPeter commented:
@Ivansete-status, @jm-clius 
Thanks for the valuable comments, I fixed those. I did not want to left for v2.
Also added some more tests. In v2 I will add even more, planned.

Ivansete-status reviewed (approved):
LGTM! Thanks for it and for your patience!
